components/script/dom/node.rs

```diff
@@ -2726,18 +2779,109 @@ impl NodeMethods for Node {
     }
 
     // https://dom.spec.whatwg.org/#dom-node-comparedocumentposition
-    fn CompareDocumentPosition(&self, other: &Node) -> u16 {
+    fn CompareDocumentPosition(&self, other: Option<&Node>) -> u16 {
         // step 1.
-        if self == other {
+        if Some(self) == other {
             return 0;
         }
 
-        // FIXME(emilio): This will eventually need to handle attribute nodes.
+        // step 2
+        let node1: Option<&Node> = other;
+        let node2: Option<&Node> = Some(self);
 
-        let mut self_and_ancestors = self
+        // step 3
+        let attr1: Option<&Attr>;
+        let attr2: Option<&Attr>;
+
+        // step 4
+        attr1 = match node1 {
+            None => None,
+            Some(n1) => match n1.downcast::<Attr>() {
+                None => None,
+                Some(attr) => Some(attr),
+            },
+        };
+
+        // slightly convoluted to ensure lifetime of attr1.GetOwnerElement
+        // outlives lifetime of node1
+        let attr1owner = match attr1 {
+            None => None,
+            Some(a) => a.GetOwnerElement(),
+        };
+
+        let node1 = match attr1owner {
+            None => node1,
+            Some(ref root1) => Some(&*root1.upcast()),
+        };
+
+        // step 5.1
+        attr2 = match node2 {
+            None => None,
+            Some(n2) => match n2.downcast::<Attr>() {
+                None => None,
+                Some(attr) => Some(attr),
+            },
+        };
+
+        let attr2owner = match attr2 {
+            None => None,
+            Some(a) => a.GetOwnerElement(),
+        };
+
+        let node2 = match attr2owner {
+            None => node2,
+            Some(ref root2) => Some(&*root2.upcast()),
+        };
+
+        // Step 5.2
+        // This substep seems lacking in test coverage.
+        if node1 == node2 && attr1 != None && attr2 != None {
```

Comment:
use `.is_some()`

Reply:
Actually this is better written as 

```rust
match (attr1, attr2) {
  (Some(a1), Some(a2)) => {
       // actual code
   }
   _ => ()
}
```

Avoid unwraps wherever possible

Reply:
I had lifetime errors with your suggested changes and ended up going back to holding the GetOwnerElement results in their own variables, but with most of your syntax suggestions.